File: ios/ShoreKit/Sources/ShoreKit/Models/Protest.swift
import Foundation
import GRDB

public struct Protest: Codable, Identifiable, Sendable {
    public var id: UUID
    public var carnivalId: UUID
    public var eventId: UUID
    public var heatId: UUID?
    public var lodgedByClubId: UUID
    public var reason: String
    public var status: String
    public var adjudicationReason: String?
    public var lodgedAt: Date
    public var adjudicatedAt: Date?

    public init(id: UUID, carnivalId: UUID, eventId: UUID, heatId: UUID? = nil, lodgedByClubId: UUID, reason: String, status: String = "Lodged", adjudicationReason: String? = nil, lodgedAt: Date = .now, adjudicatedAt: Date? = nil) {
        self.id = id
        self.carnivalId = carnivalId
        self.eventId = eventId
        self.heatId = heatId
        self.lodgedByClubId = lodgedByClubId
        self.reason = reason
        self.status = status
        self.adjudicationReason = adjudicationReason
        self.lodgedAt = lodgedAt
        self.adjudicatedAt = adjudicatedAt
    }
}

extension Protest: FetchableRecord, PersistableRecord {
    public static let databaseTableName = "protests"
}

File: src/Tides.Core/Domain/Protest.cs
using Tides.Core.Domain.Enums;

namespace Tides.Core.Domain;

public class Protest
{
    public Guid Id { get; private set; }
    public Guid CarnivalId { get; private set; }
    public Guid EventId { get; private set; }
    public Guid? HeatId { get; private set; }
    public Guid LodgedByClubId { get; private set; }
    public string Reason { get; private set; } = null!;
    public ProtestStatus Status { get; private set; }
    public string? AdjudicationReason { get; private set; }
    public DateTime LodgedAt { get; private set; }
    public DateTime? AdjudicatedAt { get; private set; }

    private Protest() { }

    public Protest(Guid id, Guid eventId, Guid? heatId, Guid lodgedByClubId, string reason)
    {
        Id = id;
        EventId = eventId;
        HeatId = heatId;
        LodgedByClubId = lodgedByClubId;
        Reason = reason;
        Status = ProtestStatus.Lodged;
        LodgedAt = DateTime.UtcNow;
    }

    public void Adjudicate(ProtestStatus outcome, string reason)
    {
        if (outcome is not (ProtestStatus.Upheld or ProtestStatus.Dismissed))
            throw new ArgumentException("Outcome must be Upheld or Dismissed.", nameof(outcome));

        Status = outcome;
        AdjudicationReason = reason;
        AdjudicatedAt = DateTime.UtcNow;
    }
}
